homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -553,104 +520,109 @@
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
         )
-        return False
-
-    hass.data[DATA_KEY][host] = device
-    async_add_entities([device], update_before_add=True)
-
-    async def async_service_handler(service):
-        """Map services to methods on XiaomiAirPurifier."""
-        method = SERVICE_TO_METHOD.get(service.service)
-        params = {
-            key: value for key, value in service.data.items() if key != ATTR_ENTITY_ID
-        }
-        entity_ids = service.data.get(ATTR_ENTITY_ID)
-        if entity_ids:
-            devices = [
-                device
-                for device in hass.data[DATA_KEY].values()
-                if device.entity_id in entity_ids
-            ]
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Fan from a config entry."""
+    entities = []
+
+    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        if DATA_KEY not in hass.data:
+            hass.data[DATA_KEY] = {}
+
+        host = config_entry.data[CONF_HOST]
+        token = config_entry.data[CONF_TOKEN]
+        name = config_entry.title
+        model = config_entry.data[CONF_MODEL]
+        unique_id = config_entry.unique_id
+
+        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+
+        if model in MODELS_PURIFIER_MIOT:
+            air_purifier = AirPurifierMiot(host, token)
+            device = XiaomiAirPurifierMiot(name, air_purifier, config_entry, unique_id)
+        elif model.startswith("zhimi.airpurifier."):
+            air_purifier = AirPurifier(host, token)
+            device = XiaomiAirPurifier(name, air_purifier, config_entry, unique_id)
+        elif model in MODELS_HUMIDIFIER_MIOT:
+            air_humidifier = AirHumidifierMiot(host, token)
+            device = XiaomiAirHumidifierMiot(
+                name, air_humidifier, config_entry, unique_id
+            )
+        elif model.startswith("zhimi.humidifier."):
+            air_humidifier = AirHumidifier(host, token, model=model)
+            device = XiaomiAirHumidifier(name, air_humidifier, config_entry, unique_id)
+        elif model.startswith("zhimi.airfresh."):
+            air_fresh = AirFresh(host, token)
+            device = XiaomiAirFresh(name, air_fresh, config_entry, unique_id)
         else:
-            devices = hass.data[DATA_KEY].values()
-
-        update_tasks = []
-        for device in devices:
-            if not hasattr(device, method["method"]):
-                continue
-            await getattr(device, method["method"])(**params)
-            update_tasks.append(device.async_update_ha_state(True))
+            _LOGGER.error(
+                "Unsupported device found! Please create an issue at "
+                "https://github.com/syssi/xiaomi_airpurifier/issues "
+                "and provide the following data: %s",
+                model,
+            )
+            return False
```

Comment:
Nothing is checking this return value.
```suggestion
            return
```